firedrake/mg/ufl_utils.py

```diff
@@ -284,6 +277,31 @@ def coarsen_snescontext(context, self, coefficient_mapping=None):
     coarse._fine = context
     context._coarse = coarse
 
+    # Now that we have the coarse snescontext, push it to the coarsened DMs
+    # Otherwise they won't have the right transfer manager when they are
+    # coarsened in turn
+    from firedrake.dmhooks import get_appctx, push_appctx, pop_appctx
+    from firedrake.dmhooks import add_hook, get_parent, SetupHooks, push_attr
+    from itertools import chain
+    for val in chain(coefficient_mapping.values(), (bc._original_val for bc in problem.bcs)):
+        if isinstance(val, firedrake.function.Function):
+            V = val.function_space()
+            coarseneddm = V.dm
+            parentdm = get_parent(coarseneddm)
+
+            # Need to make sure parentdm has a __setup_hooks__,
+            # this can happen if e.g. a BC value is assigned on Z.sub(0)
+            stack = parentdm.getAttr("__setup_hooks__")
+            if stack is None:
+                hooks = SetupHooks()
+                push_attr("__setup_hooks__", parentdm, hooks)
```

Comment:
I am concerned  that this has the potential to leave the `hooks` object dangling on the parentdm. The top-level usage of `SetupHooks` comes via the `with add_hooks` context manager which ensures that things are cleaned up.

Reply:
I think I was getting the parent from the wrong object. I think it is fixed now.